demo/admin/src/products/categories/AssignProductsGrid.tsx

```diff
@@ -0,0 +1,86 @@
+import { gql, useApolloClient, useQuery } from "@apollo/client";
+import { Savable } from "@comet/admin";
+import { CircularProgress } from "@mui/material";
+import {
+    GQLGetProductIdsForProductCategoryQuery,
+    GQLGetProductIdsForProductCategoryQueryVariables,
+    GQLSetProductCategoryMutation,
+    GQLSetProductCategoryMutationVariables,
+} from "@src/products/categories/AssignProductsGrid.generated";
+import { ProductsSelectGrid } from "@src/products/categories/ProductsSelectGrid";
+import isEqual from "lodash.isequal";
+import React, { useEffect, useMemo, useState } from "react";
+import { FormattedMessage } from "react-intl";
+
+const setProductCategoryMutation = gql`
+    mutation SetProductCategory($id: ID!, $input: [ID!]!) {
+        updateProductCategory(id: $id, input: { products: $input }) {
+            id
+        }
+    }
+`;
+
+const getProductIdsForProductCategory = gql`
+    query GetProductIdsForProductCategory($id: ID!) {
+        products(filter: { category: { equal: $id } }) {
+            nodes {
+                id
+            }
+        }
+    }
+`;
+
+interface FormProps {
+    productCategoryId: string;
+}
+
+export function AssignProductsGrid({ productCategoryId }: FormProps): React.ReactElement {
+    const client = useApolloClient();
+
+    const { data, error, loading } = useQuery<GQLGetProductIdsForProductCategoryQuery, GQLGetProductIdsForProductCategoryQueryVariables>(
+        getProductIdsForProductCategory,
+        {
+            variables: { id: productCategoryId },
+        },
+    );
+
+    const initialValues = useMemo(() => {
+        return data?.products.nodes ? data.products.nodes.map((product) => product.id) : [];
+    }, [data]);
+    const [values, setValues] = useState<string[]>(initialValues);
+    useEffect(() => {
+        setValues(initialValues);
+    }, [initialValues]);
```

Comment:
I'm pretty sure that can be done simpler. You now have 4 things that work somehow together: useQuery, useMemo, useState and useEffect. That is too complicated. (I would not care if you do that in an application, but we are creating best practice here -  that might serve as template for code generation)

Reply:
alright, I have no idea... maybe use useReducer? do you have any idea? or @johnnyomair?

Reply:
No, useReducer would be an overkill. We could probably eliminate initialValues altogether:

1. For reset, we can use data.products.nodes
2. For hasChanges, we can also use data.products.nodes. We could also add a `hasChanges` which is only updated in onChange, which would remove the need to sort the arrays in hasChanges

Reply:
I only managed to remove useMemo, it's still required to use useEffect because `data.products` is initially undefined so `values` has to be updated. (and useState can't be called after `if (loading) ...`)

Reply:
That's the best I've came up with, eliminating two out of four:

```ts
const { data, error, loading } = useQuery<GQLGetProductIdsForProductCategoryQuery, GQLGetProductIdsForProductCategoryQueryVariables>(
    getProductIdsForProductCategory,
    {
        variables: { id: productCategoryId },
        onCompleted: (data) => {
            setValues(data.products.nodes.map((product) => product.id));
        },
    },
);

const [values, setValues] = useState<string[]>(data?.products.nodes?.map((product) => product.id) ?? []);
```

What do you think?

Reply:
BTW, I'd still suggest having a hasChanges state instead of comparing and sorting two arrays at every rendern. But this is probably premature optimization

Reply:
> What do you think?

I'll take it :-D 

> BTW, I'd still suggest having a hasChanges state instead of comparing and sorting two arrays at every rendern. But this is probably premature optimization

I already added a console.log and it seems it's not called that often. probably nearly as often as the hasChange-state would change/needed to be checked. (for every click on one of the checkboxes, and additionally on closing the dialog)